### ISSUE TITLE
Error'd link with really long, suggestion-only document

### ISSUE THREAD

**Issue by patcon** [nytimes/library#103]
### Context (Environment)

- Heroku deploy
- Node 10
- `folder` type drive
- `NODE_ENV=development` on Heroku, so that no need for auth

🔗 Website: https://coronavirustechhandbook-demo.herokuapp.com/
🔢 Code: https://github.com/patcon/library-coronavirustechhandbook (no changes)
📂 Folder: https://drive.google.com/drive/folders/1EKzx34ZiyMh-3V9LoyrO3Q0SiyyAhJmE

### Expected & Actual Behavior

The deploy works fine for every world-editable doc.

I then added a suggestion-only doc ("Doctors", and it's unable to detect it -- it fails to show up on "View All" page with the rest.

When I create a _copy_ of that doc to troubleshoot ("Copy of Doctors"), it [_does_ show up in the index](https://coronavirustechhandbook-demo.herokuapp.com/medicine-care), but on clicking, it gives a 500 error, and shows this message in logs:

```
2020-03-19T23:26:54.511595+00:00 app[web.1]: info: GET /medicine-care/copy-of-doctors
2020-03-19T23:26:54.962382+00:00 app[web.1]: debug: updating tree...
2020-03-19T23:26:54.962784+00:00 app[web.1]: debug: searching for files > 0
2020-03-19T23:26:56.100738+00:00 heroku[router]: at=info method=GET path="/medicine-care/copy-of-doctors" host=coronavirustechhandbook-demo.herokuapp.com request_id=f868d030-75c9-4107-b8bf-91e0d77a0635 fwd="72.137.70.22" dyno=web.1 connect=1ms service=1594ms status=500 bytes=33839 protocol=https
2020-03-19T23:26:56.094799+00:00 app[web.1]: error: Serving an error page for /medicine-care/copy-of-doctors
2020-03-19T23:26:56.094813+00:00 app[web.1]: { message: 'Cannot read property \'displayName\' of undefined',
2020-03-19T23:26:56.094814+00:00 app[web.1]: stack:
2020-03-19T23:26:56.094817+00:00 app[web.1]: 'TypeError: Cannot read property \'displayName\' of undefined\n    at Object.exports.getProcessedDocAttributes (/app/server/formatter.js:235:52)\n    at exports.fetchDoc (/app/server/docs.js:44:57)\n    at process._tickCallback (internal/process/next_tick.js:68:7)' }
2020-03-19T23:26:56.126723+00:00 app[web.1]: debug: searching for files > 12
2020-03-19T23:26:56.509352+00:00 app[web.1]: debug: Current file count in drive: 36
2020-03-19T23:26:56.509471+00:00 app[web.1]: debug: tree updated.
```

Here's my scrubbed `.env` file: https://gist.github.com/patcon/238c9b960806e55c5e905bd576a85842

It's confusing in that creating a copy doesn't resolve it, and seems to introduce a different edge-case, that's unclear in its relation! ;)

Possible remedies:
- [x] creating a copy of file
- [x] renaming file
- [x] wholesale copying contents into new file and starting over ("Copy-paste of Doctors") (didn't work)
- [x] shortening document (didn't work)

This is admittedly a really long document (22 pages), so maybe that's an issue.

- [Doctors](https://docs.google.com/document/d/111k1L5D9TZNShV5Gr2AJ7grfuXmEx_dYXAZcNwEebQI/edit#heading=h.oi5hs7chmx59)
- [Copy of Doctors](https://docs.google.com/document/d/1MlTNiy0E4Im6VC7m6n3T4i_IE2iKsQgIpc7gHYVMgnQ/edit#)
- [Copy-paste of Doctors](https://docs.google.com/document/d/1NRkUF6oO327zCH5fWuoj1CoMRayDf_8991lopzaYKtc/edit#heading=h.a522wj6ov4bz)

**Comment by patcon** [comment]
Note that:
- docs are world editable (anyone with link can edit)
- the service account user has edit access

**Comment by patcon** [comment]
Ok, found some odd behavior:

- top level folder, owned by me (gave service account write access)
  - subdocs, owned by others, but editable with link: _work fine_
  - subdoc, owned by others, commentable with link: _fails, not shown in index_
    - copy of subdoc, owned by me, **editable** with link, which happened by default: _seen in index, but 500 error_
    - copy of subdoc, owned by me, **commentable** with link: _works!_
    - copy of subdoc, owned by me, **viewable** with link: _works!_

So basically, downgrading copied doc permissions (so that link is only viewable, not writable) allows it to be rendered in Library...!

You can see examples of all of this here, with helpfully labelled docs:
https://coronavirustechhandbook-demo.herokuapp.com/medicine-care

**Comment by isaacwhite** [comment]
Hey @patcon, could this be an intermittent issue with docs that are newly created (or moved) to your shared folder?

Because of the way Google Drive permissions propagate, we've seen some some issues where a file can take 15 minutes before it is actually readable by the service account, if the sharing permissions are derived from the folder it's in. Perhaps because different API methods are involved, the files are often listable before they are available for export (accessing the HTML).

This doesn't happen with Shared Drives, possibly because files in that context are ownerless and all reference the master permissions on the drive itself.

My hunch is that the length of the document doesn't have anything to do with it, but let me know if creating just a random document of any length you can't reproduce the behavior. The cases I'm curious which might cause the issue are:
- Create a brand new document in the folder already
- Create a brand new document outside the folder and move it in
- Take an existing document with different sharing permissions and move it into the drive

**Comment by isaacwhite** [comment]
In addition to the shared folder bug for new documents I mentioned above, I can see from your log files you may be surfacing a bug in Library around our use of revision data to determine who last changed a file.

The trace from your logs points to this line:
```javascript
const createdBy = revisionData.lastModifyingUser.displayName
```

Nearby, we have a pending TODO:
```javascript
  // TODO: guard against null revision data?
```

To work around this, I _think_ you could make any change to the copy of the document that is erroring and it would start working. I'm also going to open a PR shortly to try and resolve the underlying bug.

Thanks for reporting this!

**Comment by patcon** [comment]
Ah, I just came back to report on what I'd found.

I believe I was thrown off by a second issue related to the copies needing to be read/comment instead of write. This appears to be unrelated to the original issue of the doc not even being indexed (not showing in Library).

For the original issue, it was simply that the service account couldn't be added to someone else's doc set into comment mode 😭

**Comment by patcon** [comment]
> Hey @patcon, could this be an intermittent issue with docs that are newly created (or moved) to your shared folder?

This went on for testing over a few hours, and while I did see some small delays, this doesn't seem to explain the consistent issue I'm seeing across a few docs.

**Comment by patcon** [comment]
> The cases I'm curious which might cause the issue are:

Will try!

**Comment by patcon** [comment]
Ok, got a minimal reproduction ready:

- top-level directory is world-writable thru link. service account added (with either view-only or "organize" perms; doesn't matter)
- subdir gets 3 new files, which each inherit the service account permission bit
  - https://drive.google.com/drive/folders/1gsd3qLBhetURhn3k1wOjzGUZCe6cBLKo
  - each is set with it's link permission to be either view/suggest/write
  - write permission throws a 500 on Library

Can see this happening here:
https://coronavirustechhandbook-demo.herokuapp.com/debugging

Curious to know if this feels unexpected, and if you can guess why :) I'm down to try to fix myself with some pointers!

**Comment by isaacwhite** [comment]
Hey @patcon, I can't think of a reason that the write permission would throw, that seems very odd. Your prior log was helpful for pinpointing https://github.com/nytimes/library/pull/104.

Each time Library serves a 500 it should also print a trace in the server logs. Can you share those or can you reproduce this locally? That seems like the next step here.

**Comment by patcon** [comment]
Ok, weird.

Check this out: https://coronavirustechhandbook-demo.herokuapp.com/debugging

So it seems that a copy of the original doc ("Doctors") has some odd effects when set to "writable via link", but no other permission state for the link. (All of these direct children were created via "File > Create a copy".)

But then when I take that copy (specifically the one that's not working, due to editable link status), and in turn _copy it_, then it works fine with any permission flag set.

So there's something special about the original doc which leads its children to break under some odd edge condition (having access via link be editable), that does not transfer to _its_ children.

I have no idea what's special about the first doc...!

**Comment by patcon** [comment]
> Each time Library serves a 500 it should also print a trace in the server logs. Can you share those or can you reproduce this locally?

```
2020-03-20T01:36:56.331166+00:00 app[web.1]: info: GET /debugging/debug-copy-of-existing-doc-writable-borked
2020-03-20T01:36:57.812280+00:00 app[web.1]: error: Serving an error page for /debugging/debug-copy-of-existing-doc-writable-borked
2020-03-20T01:36:57.812296+00:00 app[web.1]: { message: 'Cannot read property \'displayName\' of undefined',
2020-03-20T01:36:57.812297+00:00 app[web.1]: stack:
2020-03-20T01:36:57.812308+00:00 app[web.1]: 'TypeError: Cannot read property \'displayName\' of undefined\n    at Object.exports.getProcessedDocAttributes (/app/server/formatter.js:235:52)\n    at exports.fetchDoc (/app/server/docs.js:44:57)\n    at process._tickCallback (internal/process/next_tick.js:68:7)' }
2020-03-20T01:36:57.819619+00:00 heroku[router]: at=info method=GET path="/debugging/debug-copy-of-existing-doc-writable-borked" host=coronavirustechhandbook-demo.herokuapp.com request_id=3e442384-b1ae-4943-84e4-a5a7f0ef0ba7 fwd="72.137.70.22" dyno=web.1 connect=1ms service=1491ms status=500 bytes=33839 protocol=https
2020-03-20T01:36:57.936797+00:00 heroku[router]: at=info method=GET path="/whoami.json" host=coronavirustechhandbook-demo.herokuapp.com request_id=585a3c5e-fb0f-44b4-97af-6c36e20b138a fwd="72.137.70.22" dyno=web.1 connect=1ms service=2ms status=304 bytes=562 protocol=https
2020-03-20T01:37:01.517910+00:00 app[web.1]: debug: updating tree...
2020-03-20T01:37:01.518515+00:00 app[web.1]: debug: searching for files > 0
2020-03-20T01:37:01.774439+00:00 app[web.1]: debug: searching for files > 12
2020-03-20T01:37:02.903297+00:00 app[web.1]: debug: Current file count in drive: 43
2020-03-20T01:37:02.903546+00:00 app[web.1]: debug: tree updated.
```

Here ya go!

**Comment by isaacwhite** [comment]
👍 if those are the only errors in the logs, the PR we just merged ought to fix it. The problem there is Library is receiving empty revision data, which previously did not have an adequate null check. Fixed here: https://github.com/nytimes/library/pull/104

**Comment by patcon** [comment]
Ok, just tested the new master locally, and seems to work fine for that ;)

Thanks so much @isaacwhite! You def helped me untangle the bug from my own goofs 😄

**Comment by isaacwhite** [comment]
Happy to help! 

If you want to customize your site further, definitely check out [our customization](https://github.com/nytimes/library-customization-example) example, which powers https://nyt-library-demo.herokuapp.com/
